### ISSUE TITLE
[FEATURE] Add option to raise exception if snapshot doesn't complete with SUCCESS

### ISSUE THREAD

**Issue by kellycampbell** [elastic/curator#1192]
I have a curator config with two actions: 1. snapshot, 2. delete_indices.

I've set `continue_with_exception` to `False` in step 1, but it is still continuing and deleting indices.

## Log output
```
2018-04-02 19:08:41,143 INFO      Preparing Action ID: 1, "snapshot"
2018-04-02 19:08:41,148 INFO      Trying Action ID: 1, "snapshot": Archive indexes older than 1 days
2018-04-02 19:08:42,231 INFO      Creating snapshot "curator-20180402190841" from indices: [ u'logstash-2018.03.23', u'logstash-2018.03.22', u'logstash-2018.03.21', u'logstash-2018.03.20']
2018-04-02 19:08:48,022 INFO      Snapshot curator-20180402190841 still in progress.
2018-04-02 19:09:08,200 ERROR     Snapshot curator-20180402190841 completed with state FAILED.
2018-04-02 19:09:08,200 INFO      Action ID: 1, "snapshot" completed.
2018-04-02 19:09:08,200 INFO      Preparing Action ID: 2, "delete_indices"
2018-04-02 19:09:08,202 INFO      Trying Action ID: 2, "delete_indices": Clean up ES by deleting old indices
2018-04-02 19:09:08,972 INFO      Deleting selected indices:  ... <cut>
```

My config is the following:

```
actions:
  1:
    action: snapshot
    description: "Archive indexes older than 1 days"
    options:
      repository: s3-archive
      wait_for_completion: True
      max_wait: 3600
      wait_interval: 20
      continue_if_exception: False
    filters:
    - filtertype: age
      source: name
      direction: older
      timestring: '%Y.%m.%d'
      unit: days
      unit_count: 1
  2:
    action: delete_indices
    description: "Clean up ES by deleting old indices"
    options:
      timeout_override:
      continue_if_exception: True
      disable_action: False
    filters:
    - filtertype: age
      source: name
      direction: older
      timestring: '%Y.%m.%d'
      unit: days
      unit_count: 7
      field:
      stats_result:
      epoch:
      exclude: False
```

### Expected Behavior

It should not continue to the delete action.

### Actual Behavior

It deletes the indices even though they haven't been archived yet.

### Specifications

  - Version: curator 5.4.0 (hub.docker.com/bobrik/curator:latest
  - ES Version: 6.1.1
  - Platform: linux

**Comment by untergeek** [comment]
This represents a slight misunderstanding of what happens under the hood.  A snapshot completing with state `FAILED` can still represent successful execution of the Snapshot API—which is exactly what happened, based on the log lines you provided.  In other words, `continue_if_exception` only catches code _exceptions,_ and one did not happen here.

Snapshot failure is never a good thing.  Do you know what's going on with this?  Have you checked the Elasticsearch logs to see why it's failing?  That might be a more pressing concern than altering this feature in Curator, as you're not even getting a `PARTIAL` snapshot state—it's outright saying the snapshot has `FAILED`. (See more on snapshot status codes [here](https://www.elastic.co/guide/en/elasticsearch/reference/6.2/modules-snapshots.html#_snapshot)).

With that said, it might be a good feature to add an option flag to force Curator to _raise_ an exception if a snapshot doesn't complete with state `SUCCESS`, for example.

**Comment by kellycampbell** [comment]
I think it was something thing with our AWS s3 credentials. I guess I
expected a failure like that to cause an exception. In this case it's a
data loss without that exception. Fortunately this was on a test cluster so
not a major loss.

**Comment by untergeek** [comment]
Yeah, it's not a failure if the API call returns successfully, regardless of what state the snapshot is in.  I do like the idea of setting a flag to raise an exception if the state is something other than `SUCCESS`.  I will change this to a feature request ticket.

**Comment by kellycampbell** [comment]
Just curious, will this new flag be defaulted to true?

Perhaps I'm not aware of the use-cases you are supporting by considering a failed snapshot as non-exceptional, but it seems like raising the exception should be the expected behavior, even though different from current behavior.

**Comment by untergeek** [comment]
A fair question.  In Curator 5.x (should I add it before releasing 6.x), it would default to not change the current behavior.  I could perhaps change it to raise an exception by default in 6.x.

**Comment by untergeek** [comment]
As a reason, I would offer that you are the first person to whom this has happened, or at least reported it and considered it a fault of Curator.

**Comment by kellycampbell** [comment]
Yes, that is a bit surprising. I'm using it via kubernetes, and I am on an older version of that which doesn't yet support periodically scheduled jobs, so I've only been running curator manually about once a month. I have noticed that I've had to recreate my s3 snapshot in ES sometimes, and that was the case when it failed yesterday.

I think the "principle of least astonishment" might be a good guide here for determining the default behavior of this new flag in 6.x.

**Comment by alexlance** [comment]
I have experienced this recently too. It was alarming to discover that the snapshots were only partial snapshots. It would be excellent if curator threw an error on a failed or partial snapshot.

**Comment by lucas044** [comment]
Yes this is a huge problem for me, a PARTIAL snapshot is not really useful to anyone. Would love some hard failures, curator deleted all my older snapshots automatically (as per retention configuration) so any last SUCCESS snapshots were removed.

**Comment by untergeek** [comment]
This is fixed as of #1221 and fixed in release 5.5.3.
